### DISCUSSION TITLE
Why clangd is choosing header of wrong version

### DISCUSSION THREAD

**Discussion by Masquue** [clangd/clangd#1412] [Q&A] [answered]
I have both `g++-11` and `g++-12` installed, and have `g++` pointing to `g++-11`.
`compile_commands.json` is correctly using `/usr/bin/g++`.
But the parse error "\<iostream\> not found" (I did not install `g++-12` in the beginning) and the "--log=verbose" output both shows that `clangd` is always trying to find headers in the directories of version 12.

The "Heuristic search for system headers" section in the `clangd` website did not answer my question.
I wonder why this is the case.

**Comment by HighCommander4** [comment, marked as answer]
This is an issue others have run into as well.

I don't have a full answer, but I've written down my best guess, along with possible workarounds, here:

https://github.com/clangd/clangd/issues/1394#issuecomment-1328676884

**Reply by Masquue** [reply]
This is exactly my situation. Somehow I had `gcc-12` installed but not `g++-12` in my ubuntu 22.04 VM. 
And after installing it, the problem resolved.

Still strange why `clangd` is not using the compiler in the `compile_commands.json` to find header location,
even if I explicitly use `/usr/bin/g++-11` in `compile_commands.json`, it is still using version 12 headers.
I feel this is a huge mistake.

But after all my problem is solved. And thank you very much.

**Reply by HighCommander4** [reply]
Clangd does not **execute** the compiler in the `compile_commands.json` (to ask it for its built-in include paths) unless `--query-driver` is used (see "option 3" from my linked comment). For why `--query-driver` is not the default, see https://github.com/clangd/clangd/issues/539.

**Reply by Masquue** [reply]
Totally agree that security is important factor to consider.
It is sad that `--query-driver` is not a per-project setting (the `--query-driver=/**/*` solution loses its security).
But it seems that `--query-driver`'s value can be a list of compilers? However I haven't found it mentioned in other places except this [comment](https://github.com/clangd/clangd/issues/537#issuecomment-696551532).

Also agree maybe a better solution is to improve the heuristic algorithm to take into account of the compiler
in the `compile_commands.json` to decide the right header location, or in this case, the right version.

**Reply by HighCommander4** [reply]
> It is sad that `--query-driver` is not a per-project setting

Depending on your editor, you may be able to specify command-line arguments to clangd on a per-project basis. For example, with vscode, you specify `"clangd.arguments"` in `<project>/.vscode/settings.json`.

> But it seems that `--query-driver`'s value can be a list of compilers? However I haven't found it mentioned in other places except this [comment](https://github.com/clangd/clangd/issues/537#issuecomment-696551532).

Please run `clangd --help` and see the description of the `--query-driver` option; it does mention both the ability to specify multiple paths, and the ability to use globs.

> Also agree maybe a better solution is to improve the heuristic algorithm to take into account of the compiler in the `compile_commands.json` to decide the right header location, or in this case, the right version.

I agree; I mentioned this at the end of my comment in the other issue:

> I think a proper fix would involve fixing the clang driver's heuristics for locating standard library paths, to handle this situation correctly. However, the code in question is not in clangd, it's upstream in libclangDriver, and I don't have the bandwidth to investigate it further at this time.
